Add cpo elastic tensor plugin

MFraters opened the pull request:
This  pull request add a particle property plugin to compute the full elastic tensor for the CPO particle properties. It is dependent on #4975, so for now I have simplified it a bit to remove the dependency until #4975 is merged. 

This plugin contains a lot of utility functions which may be of general use (all the tensor transform functions). I don't thing they belong in this class, but I am wondering where they would fist best, an internal utilities namespace, or the general utilities file or some of them maybe in deal.ii?

I am not very happy with the hardcoded stiffmatrices for olivine and enstatite. Does it make sense to make them input parameters. It seems thateven though they are second order tensors with a dimension 6, they only have 9 unique values (at least for olivine and enstatite).

(point 6 in #3885)

MFraters commented:
This pull request is now ready for review.

gassmoeller commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,582 @@
+            for (unsigned short int k = 0; k < 3; k++)
+              for (unsigned short int l = 0; l < 3; l++)
+                {
+                  // The first part of the inline if statment gets the diagonal.
```
```suggestion
                  // The first part of the inline if statement gets the diagonal.
```

gassmoeller commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,582 @@
+              for (unsigned short int l = 0; l < 3; l++)
+                {
+                  // The first part of the inline if statment gets the diagonal.
+                  // The second part is never higher then 5 (which is the limit of the tensor index)
```
```suggestion
                  // The second part is never higher than 5 (which is the limit of the tensor index)
```

gassmoeller commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,582 @@
+      {
+        SymmetricTensor<2,6> result;
+
+        constexpr double sqrt_2_inv = 1/sqrt(2);
```
If you are concerned about speed you should do the same in the previous function with `sqrt(2)`. There is already a deal.II number in `numbers::SQRT2` that you can use here and above.

gassmoeller commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,582 @@
+      void
+      CpoElasticTensor<dim>::declare_parameters (ParameterHandler &prm)
+      {
+        prm.enter_subsection("Postprocess");
+        {
+          prm.enter_subsection("Particles");
+          {
+            prm.enter_subsection("Crystal Preferred Orientation");
+            {
+              prm.declare_entry ("Number of grains per particle", "50",
+                                 Patterns::Integer (1),
+                                 "The number of grains of each different mineral "
+                                 "each particle contains.");
+
+              prm.enter_subsection("Initial grains");
+              {
+                prm.enter_subsection("Uniform grains and random uniform rotations");
+                {
+                  prm.declare_entry ("Minerals", "Olivine: Karato 2008, Enstatite",
+                                     Patterns::List(Patterns::Anything()),
+                                     "This determines what minerals and fabrics or fabric selectors are used used for the LPO calculation. "
+                                     "The options are Olivine: Passive, A-fabric, Olivine: B-fabric, Olivine: C-fabric, Olivine: D-fabric, "
+                                     "Olivine: E-fabric, Olivine: Karato 2008 or Enstatite. Passive sets all RRSS entries to the maximum. The "
+                                     "Karato 2008 selector selects a fabric based on stress and water content as defined in "
+                                     "figure 4 of the Karato 2008 review paper (doi: 10.1146/annurev.earth.36.031207.124120).");
+                }
+                prm.leave_subsection ();
+              }
+              prm.leave_subsection ();
+            }
+            prm.leave_subsection ();
+          }
+          prm.leave_subsection ();
+        }
+        prm.leave_subsection ();
+      }
```
Arent all of these parameters already declared in the main  CPO plugin? No need to declare them again here. Just keep the parse_parameters function and leave a comment that describes where the parameters have been declared. Using only a single declaration is safer in case we decide to rename the variables.

gassmoeller commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,582 @@
+      ASPECT_REGISTER_PARTICLE_PROPERTY(CpoElasticTensor,
+                                        "cpo elastic tensor",
+                                        "A plugin in which the particle property tensor is "
+                                        "defined as the deformation the Voigt average of the "
```
```suggestion
                                        "defined as the Voigt average of the "
```

gassmoeller reviewed (commented):
Sorry this took so long. I have no big concerns about this, my only major question is if all of these conversion functions are needed. Do you need them for a follow-up PR? Do they need to be static or could they be utility functions outside of the class?

bobmyhill commented on include/aspect/particle/property/cpo_elastic_tensor.h:
```diff
@@ -0,0 +1,246 @@
+           * Rotate a 3D 4th order tensor with an other 3D 4th
+           */
+          static
+          Tensor<4,3> rotate_4th_order_tensor(const Tensor<4,3> &input_tensor, const Tensor<2,3> &rotation_tensor);
```
For this function and the others, you should be able to use the deal.II symmetric tensor (then you only need to specify 36 elements)

bobmyhill reviewed (changes requested):
This one is good for the Voigt vs. Mandel forms [doi](https://en.wikipedia.org/wiki/Doi_(identifier)):[10.1016/s0045-7825(00)00263-2](https://doi.org/10.1016%2Fs0045-7825%2800%2900263-2).

We should stick with Voigt, I think, and double check that we use the resulting 1x6 strain tensor consistently with that choice.

MFraters commented:
I have rebased and fixed the code with the new tensor utilities functions.

bobmyhill commented on include/aspect/particle/property/cpo_elastic_tensor.h:
```diff
@@ -0,0 +1,199 @@
+       * Computes the elastic tensor $C_{ijkl} based on the cpo in both olivine
+       * and enstatite. It uses a Voigt average.
+       *
+       * The layout of the data vector per partcle is the following (note that
```
```suggestion
       * The layout of the data vector per particle is the following (note that
```

bobmyhill commented on include/aspect/particle/property/cpo_elastic_tensor.h:
```diff
@@ -0,0 +1,199 @@
+       * and enstatite. It uses a Voigt average.
+       *
+       * The layout of the data vector per partcle is the following (note that
+       * for this plugin the following dim's are always 3):
```
```suggestion
       * for this plugin the following dims are always 3):
```

bobmyhill commented on include/aspect/particle/property/cpo_elastic_tensor.h:
```diff
@@ -0,0 +1,199 @@
+       *
+       * This class contains functions to convert the elastic tensor in
+       * different forms: Tensor<4,3>, Tensor<2,6>, Tensor<1,21>. These
+       * forms are all equivilant, but are useful for different operations.
```
```suggestion
       * forms are all equivalent, but are useful for different operations.
```

bobmyhill commented on include/aspect/particle/property/cpo_elastic_tensor.h:
```diff
@@ -0,0 +1,199 @@
+
+          /**
+           * Update function. This function is called every time an update is
+           * request by need_update() for every particle for every property.
```
```suggestion
           * requested by need_update() for every particle for every property.
```

bobmyhill commented on include/aspect/particle/property/cpo_elastic_tensor.h:
```diff
@@ -0,0 +1,199 @@
+                                        const ArrayView<double> &particle_properties) const override;
+
+          /**
+           * This implementation tells the particle manager that
```
```suggestion
           * This function tells the particle manager that
```

bobmyhill commented on include/aspect/particle/property/cpo_elastic_tensor.h:
```diff
@@ -0,0 +1,199 @@
+
+          /**
+           * Return which data has to be provided to update the property.
+           * The integrated strains needs the gradients of the velocity.
```
```suggestion
           * For example, the integrated strains needs the gradients of the velocity.
```

bobmyhill commented on include/aspect/particle/property/cpo_elastic_tensor.h:
```diff
@@ -0,0 +1,199 @@
+          get_property_information() const override;
+
+          /**
+           * Loads particle data into variables
```
Does it? In that case this function name could be better.

bobmyhill commented on include/aspect/particle/property/cpo_elastic_tensor.h:
```diff
@@ -0,0 +1,199 @@
+                             const ArrayView<double> &data);
+
+          /**
+           * Stores information in variables into the data array
```
```suggestion
           * Populates data array with provided elastic tensor values at a specific data position.
```

bobmyhill commented on include/aspect/particle/property/cpo_elastic_tensor.h:
```diff
@@ -0,0 +1,199 @@
+          declare_parameters (ParameterHandler &prm);
+
+          /**
+           * Read the parameters this class declares from the parameter file.
```
```suggestion
           * Parses the parameters this class declares from the parameter file.
```

bobmyhill commented on include/aspect/particle/property/cpo_elastic_tensor.h:
```diff
@@ -0,0 +1,199 @@
+        private:
+          unsigned int cpo_data_position;
+
+          SymmetricTensor<2,6> stiffness_matrix_olivine;
```
Is this an appropriate time to generalise to std::vector<SymmetricTensor<2,6>> stiffness_matrices?

I realise this might be a lot of work, so an alternative might be to rename this .h and .cc files and the class to better describe what they do. Maybe `OlivineEnstatiteCPO`?

MFraters replied:
as discussed, let's leave the name for now and generatlize it in a follow up pull request.

bobmyhill commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,313 @@
+      template <int dim>
+      CpoElasticTensor<dim>::CpoElasticTensor ()
+      {
+        permutation_operator_3d[0][1][2]  = 1;
```
I think it would be useful to make a standard Levi-Civita tensor somewhere.

MFraters replied:
See #5318

bobmyhill commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,313 @@
+        permutation_operator_3d[2][1][0]  = -1;
+
+        // The following values are directly from D-Rex.
+        // Todo: make them a input parameter
```
Yes, again, for generalisation this would be good.

bobmyhill commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,313 @@
+               ExcMessage("No s wave anisotropy property plugin found."));
+
+        AssertThrow(manager.check_plugin_order("crystal preferred orientation","cpo elastic tensor"),
+                    ExcMessage("To use the cpo elastic tensor plugin, the cpo plugin need to be defined before this plugin."));
```
```suggestion
                    ExcMessage("To use the cpo elastic tensor plugin, the cpo plugin needs to be defined before this plugin."));
```

bobmyhill commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,313 @@
+                {
+                  prm.declare_entry ("Minerals", "Olivine: Karato 2008, Enstatite",
+                                     Patterns::List(Patterns::Anything()),
+                                     "This determines what minerals and fabrics or fabric selectors are used used for the LPO calculation. "
```
```suggestion
                                     "This determines what minerals and fabrics or fabric selectors are used for the LPO calculation. "
```

bobmyhill commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,313 @@
+                                     "The options are Olivine: Passive, A-fabric, Olivine: B-fabric, Olivine: C-fabric, Olivine: D-fabric, "
+                                     "Olivine: E-fabric, Olivine: Karato 2008 or Enstatite. Passive sets all RRSS entries to the maximum. The "
+                                     "Karato 2008 selector selects a fabric based on stress and water content as defined in "
+                                     "figure 4 of the Karato 2008 review paper (doi: 10.1146/annurev.earth.36.031207.124120).");
```
```suggestion
                                     "Figure 4 of the Karato 2008 review paper (doi: 10.1146/annurev.earth.36.031207.124120).");
```

bobmyhill commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,313 @@
+                                        "cpo elastic tensor",
+                                        "A plugin in which the particle property tensor is "
+                                        "defined as the Voigt average of the "
+                                        "elastic tensors of the cpo grains of the minerals."
```
```suggestion
                                        "elastic tensors of the minerals in the textured rock."
```

bobmyhill commented on unit_tests/crystal_preferred_orientation.cc:
```diff
@@ -1067,3 +1064,389 @@ TEST_CASE("CPO")
+                               37
+                              };
+
+  // There used be be 36 unique entries, but now because we are using the
```
```suggestion
  // There used to be 36 unique entries, but now because we are using the
```

bobmyhill reviewed (changes requested):
Some more comments, and @gassmoeller has some comments that haven't yet been addressed either.

MFraters commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,582 @@
+                                                           const ArrayView<double> &data) const
+      {
+        SymmetricTensor<2,6,double> Sav;
+        const SymmetricTensor<2,6> *stiffness_matrix = &stiffness_matrix_olivine;
```
as discussed, we keep this for now. This could be changed into a lambda function in the future.

MFraters commented:
I have addressed all the comments (I reopened all resolved conversations to check). It is ready for another review.

gassmoeller commented:
Unit tests fail ;-)

MFraters commented:
Fixed!

(at least on my computer)

gassmoeller commented on include/aspect/particle/property/cpo_elastic_tensor.h:
```diff
@@ -0,0 +1,204 @@
+/*
+ Copyright (C) 2022 by the authors of the ASPECT code.
```
2023

gassmoeller commented on include/aspect/particle/property/cpo_elastic_tensor.h:
```diff
@@ -0,0 +1,204 @@
+       * Computes the elastic tensor $C_{ijkl} based on the cpo in both olivine
+       * and enstatite. It uses a Voigt average.
+       *
+       * The layout of the data vector per partcle is the following (note that
```
particle

gassmoeller commented on include/aspect/particle/property/cpo_elastic_tensor.h:
```diff
@@ -0,0 +1,204 @@
+           */
+          void
+          initialize () override;
+
```
one empty line

gassmoeller commented on include/aspect/particle/property/cpo_elastic_tensor.h:
```diff
@@ -0,0 +1,204 @@
+           * Mainprice (1990): $C_{ijkl} = R_{ip} R_{jg} R_{kr} R_{is} C_{pgrs}$, where
+           * R_{ij} is the cpo orientation matrix.
+           */
+          virtual
```
Does this function have to be virtual? Does it have to be public? If not, make it not virtual or not public.

MFraters replied:
It doesn't have to be virtual, but it does have to be public. We could maybe move it to the new Utilities tensor space in a follow up pull request.

gassmoeller commented on include/aspect/particle/property/cpo_elastic_tensor.h:
```diff
@@ -0,0 +1,204 @@
+
+          /**
+           * Update function. This function is called every time an update is
+           * request by need_update() for every particle for every property.
```
requested

gassmoeller commented on include/aspect/particle/property/cpo_elastic_tensor.h:
```diff
@@ -0,0 +1,204 @@
+
+          /**
+           * The stiffness tensors for olivine and enstatite.
+           * Todo: generatize this into a vector.
```
```suggestion
           * Todo: generalize this into a vector.
```

gassmoeller commented on include/aspect/particle/property/cpo_elastic_tensor.h:
```diff
@@ -0,0 +1,204 @@
+          SymmetricTensor<2,6> stiffness_matrix_enstatite;
+
+          /**
+           * The number of grain per particle
```
```suggestion
           * The number of grains per particle.
```

gassmoeller commented on include/aspect/particle/property/cpo_elastic_tensor.h:
```diff
@@ -0,0 +1,204 @@
+          unsigned int n_grains;
+
+          /**
+           * The number of minerals per particle
```
```suggestion
           * The number of minerals per particle.
```

gassmoeller commented on source/particle/property/crystal_preferred_orientation.cc:
```diff
@@ -787,7 +780,7 @@ namespace aspect
             if (volume_fraction_grain >= threshold_GBS/n_grains)
               {
-                deriv_a_cosine_matrices[grain_i] = permutation_operator_3d * w  * nondimensionalization_value;
+                deriv_a_cosine_matrices[grain_i] = Utilities::Tensors::levi_civita<3> *w  * nondimensionalization_value;
```
```suggestion
                deriv_a_cosine_matrices[grain_i] = Utilities::Tensors::levi_civita<3> * w  * nondimensionalization_value;
```

gassmoeller commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,273 @@
+                    ExcMessage("No crystal preferred orientation property plugin found."));
+
+        AssertThrow(manager.check_plugin_order("crystal preferred orientation","cpo elastic tensor"),
+                    ExcMessage("To use the cpo elastic tensor plugin, the cpo plugin need to be defined before this plugin."));
```
needs
```suggestion
                    ExcMessage("To use the cpo elastic tensor plugin, the cpo plugin needs to be defined before this plugin."));
```

gassmoeller commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,582 @@
+                                                           const unsigned int cpo_data_position,
+                                                           const ArrayView<double> &data) const
+      {
+        SymmetricTensor<2,6,double> Sav;
```
Why S and not C?
```suggestion
        SymmetricTensor<2,6,double> S_average;
```

gassmoeller commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,273 @@
+          this->get_particle_world().get_property_manager().template get_matching_property<Particle::Property::CrystalPreferredOrientation<dim>>();
+
+
+        SymmetricTensor<2,6> C_average = voigt_average_elastic_tensor(cpo_particle_property,
```
```suggestion
        const SymmetricTensor<2,6> C_average = voigt_average_elastic_tensor(cpo_particle_property,
```

gassmoeller commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,273 @@
+      {
+        return update_output_step;
+      }
+
```
three empty lines

gassmoeller commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,273 @@
+      {
+        return update_default;
+      }
+
```
as above

gassmoeller commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,273 @@
+
+        return property_information;
+      }
+
```
as above

gassmoeller commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,273 @@
+      void
+      CpoElasticTensor<dim>::declare_parameters (ParameterHandler &)
+      {}
+
```
as above

gassmoeller commented on source/particle/property/cpo_elastic_tensor.cc:
```diff
@@ -0,0 +1,273 @@
+          prm.leave_subsection ();
+        }
+        prm.leave_subsection ();
+
```
Remove one or both of these empty lines, it looks unusual to have two empy lines at the end of a function but before the closing brace.

gassmoeller commented on unit_tests/crystal_preferred_orientation.cc:
```diff
@@ -1067,3 +1064,389 @@ TEST_CASE("CPO")
+
+
+  cpo_elastic_tensor.parse_parameters(prm);
+  //cpo_elastic_tensor.initialize();
```
remove commented line

gassmoeller reviewed (commented):
Looks pretty good. Just some typos. Also check resolved conversations, I commented on some, but I dont know if they become unresolved when I do that.

MFraters commented:
oke, I now should have really addressed all comments. My apologies to @bobmyhill to accidentally removing all his suggestions that I applied, probably with a force push...

gassmoeller reviewed (approved):
Thanks. Good to go when the testers are happy.

gassmoeller commented:
@bobmyhill you also need to approve, otherwise I cannot merge.

bobmyhill commented:
Thanks :) 
Good to go when the testers are finished.
